src/Microsoft.AspNet.Hosting/Startup/StartupErrorUtilities.cs

```diff
@@ -0,0 +1,537 @@
+        private static readonly string _compilationExceptionFormatString = GetResourceString("Compilation_Exception.html");
+
+        public static byte[] GenerateErrorHtml(bool showDetails, IRuntimeEnvironment runtimeEnvironment, params object[] errorDetails)
+        {
+            if (!showDetails)
+            {
+                errorDetails = new[] { "An error occurred while starting the application." };
+            }
+
+            // Build the message for each error
+            var wasSourceCodeWrittenOntoPage = false;
+            var builder = new StringBuilder();
+            var rawExceptionDetails = new StringBuilder();
+
+            foreach (object error in errorDetails ?? new object[0])
+            {
+                Exception ex = error as Exception;
+                if (ex == null && error is ExceptionDispatchInfo)
+                {
+                    ex = ((ExceptionDispatchInfo)error).SourceException;
+                }
+
+                if (ex != null)
+                {
+                    var flattenedExceptions = FlattenAndReverseExceptionTree(ex);
+
+                    var compilationException = flattenedExceptions.OfType<ICompilationException>()
+                                                                  .FirstOrDefault();
+                    if (compilationException != null)
+                    {
+                        WriteException(compilationException, builder, ref wasSourceCodeWrittenOntoPage);
+
+                        var compilationErrorMessages = compilationException.CompilationFailures
+                            .SelectMany(f => f.Messages.Select(m => m.FormattedMessage))
+                            .Take(MaxCompilationErrorsToShow);
+
+                        WriteRawExceptionDetails("Show raw compilation error details", compilationErrorMessages, rawExceptionDetails);
+                    }
+                    else
+                    {
+                        foreach (var innerEx in flattenedExceptions)
+                        {
+                            WriteException(innerEx, builder, ref wasSourceCodeWrittenOntoPage);
+                        }
+
+                        WriteRawExceptionDetails("Show raw exception details", new[] { ex.ToString() }, rawExceptionDetails);
+                    }
+                }
+                else
+                {
+                    string message = Convert.ToString(error, CultureInfo.InvariantCulture);
+                    WriteMessage(message, builder);
+                }
+            }
+
+            // Generate the footer
+            string footer = showDetails ? GenerateFooterEncoded(runtimeEnvironment) : null;
+
+            // And generate the full markup
+            return Encoding.UTF8.GetBytes(String.Format(CultureInfo.InvariantCulture, _errorPageFormatString, builder, rawExceptionDetails, footer));
+        }
+
+
+#if !DNXCORE50
```

Comment:
They added StackFrame to CoreCLR https://www.nuget.org/packages/System.Diagnostics.StackTrace/4.0.1-beta-23225